Test bundle, Documentation & Codi 🐶 Update

RobAndrewHurst opened the pull request:
## Description

Implemented a comprehensive test bundle system to enhance test reusability across different environments. This update includes:

### Test Bundle Features

- Created reusable workspace assets for cross-environment testing
- Did a review and optimization of the majority of existing test suites. (Some still need to be revisted and reworked.)
- Standardized test configurations for better maintainability

### Codi 🐶 v0.0.47 Updates

- Enhanced test output control with `--quiet` flag
  - Streamlines output by displaying only test failures
  - Reduces noise in CI/CD pipelines
  - Backward compatible with existing test configurations

## GitHub Issue

Resolves #1440

## Type of Change

- ✅ New feature (non-breaking change which adds functionality)
  - Test bundle system
  - Enhanced output control
- ✅ Documentation
  - Updated test configuration guides
- ✅ Testing
  - Improved test organization
  - Enhanced cross-environment compatibility
  - Optimized test asset reusability

RobAndrewHurst commented on esbuild.config.mjs:
```diff
@@ -4,13 +4,14 @@ import * as esbuild from 'esbuild'
 
 const buildOptions = {
-    entryPoints: ['./lib/mapp.mjs', './lib/ui.mjs'],
+    entryPoints: isDev ? ['./lib/mapp.mjs', './lib/ui.mjs'] : ['./lib/mapp.mjs', './lib/ui.mjs', './tests/_mapp.test.mjs'],
```
We bundle the test assets if we are in DEVELOPMENT mode. We can discuss if we want this all the time or in other certain situations.

RobAndrewHurst commented on package.json:
```diff
@@ -11,7 +11,7 @@
     "test": "codi tests",
     "_build": "node esbuild.config.mjs",
-    "browser-sync": "npx browser-sync start --proxy localhost:3000 --port 3001 --ui-port 3002 --files public/js/lib/**/* --no-open --no-notify",
+    "browser-sync": "npx browser-sync start --proxy localhost:3000 --port 3001 --ui-port 3002 --files public/js/**/* --no-open --no-notify",
```
Browser sync will now reload for either changes to lib or tests

RobAndrewHurst commented on public/views/_test.html:
```diff
@@ -353,7 +344,13 @@
+            if (integrity) {
+                results = await codi.runWebTests(testFiles);
+            } else {
+                results = await codi.runWebTestFunction(_mappTest.coreTest);
```
This webTestFunction might be better used on each test function in the test bundle to try aid parallel testing in the future.

RobAndrewHurst commented on tests/_mapp.test.mjs:
```diff
@@ -0,0 +1,39 @@
+import { formatTest } from './lib/layer/format/_format.test.mjs';
+import { ui_locations } from './lib/ui/locations/_locations.test.mjs';
+
+self._mappTest = {
```
This is the test bundle entry point that self assigns just like mapp

RobAndrewHurst commented on tests/assets/layers/cluster/layer.json:
```diff
@@ -0,0 +1,67 @@
+{
```
Example of saving away a configuration, which can be imported from the bundled test library.

RobAndrewHurst commented on tests/lib/layer/format/mvt.test.mjs:
```diff
@@ -1,167 +1,33 @@
-                    }
-                }
-            }
+import mvt_layer_default from '../../../assets/layers/mvt/layer.json';
```
Example of importing the asset. This will be applied more as I continue the re-review and structure of current tests

RobAndrewHurst commented on tests/lib/mapview/allfeatures.test.mjs:
```diff
@@ -1,7 +1,7 @@
-            //TODO
-        });
-    });
+    // await codi.describe('TODO: Mapview: allfeaturesTest', async () => {
```
Keeping these TODO's in the test base so I can get to the asap

AlexanderGeere reviewed (approved):
The quiet flag is a nice addition. The testing.md makes sense to me 👍

simon-leech reviewed (changes requested):
The integrity tests are not working correctly. 

At the start of the integrity test it throws an error that /js/tests/_mapp.test.js does not exist (404). 

I also see an error 'Erroringrunning test file' using this path /public/tests/integrity.test.mjs. This is also timing out after 5000ms in the webRunner. 

The tests then also say that they have finished way before all the thematics have finished running through the test.

dbauszus-glx reviewed (dismissed):
The nodemon.json references local resources and must be added to ignore. The file and it's content must be referenced in the testing.md to explain the build process.

RobAndrewHurst commented:
> The nodemon.json references local resources and must be added to ignore. The file and it's content must be referenced in the testing.md to explain the build process.

I have updated the documentation and removed nodemon.json and added in an example config into the DEVELOPING docs

RobAndrewHurst commented:
After some discussion @simon-leech has pushed a change where the test bundle will happen all the time. This is fine because only on the _test.html view do we call the bundle!
